mirrord/layer/src/exec.rs

```diff
@@ -96,108 +95,66 @@ fn intercept_tmp_dir(argv_arr: &Nul<*const c_char>) -> Detour<Vec<*const c_char>
-            Self {
-                original_path,
-                path_c_string,
-                original_argv,
-                argv_vec,
-            }
-        }
-
-        /// Get the (maybe patched) path that should be executed.
-        pub fn get_path(&self) -> *const c_char {
-            // If we created a new string return a pointer to that, otherwise just the original
-            // pointer.
-            self.path_c_string
-                .as_ref()
-                .map(|path| path.as_ptr())
-                .unwrap_or(self.original_path)
-        }
 
-        /// Get the (maybe changed) argv.
-        pub fn get_argv(&self) -> *const *const c_char {
-            // If we created a new array return a pointer to that, otherwise just the original
-            // pointer.
-            self.argv_vec
-                .as_ref()
-                .map(|argv| argv.as_ptr())
-                .unwrap_or(self.original_argv)
-        }
+        c_string_vec.push(
+            CString::new(arg_str
+                .strip_prefix(&tmp_dir)
+                .inspect(|original_path| {
+                    trace!(
+                        "Intercepted mirrord's temp dir in argv: {}. Replacing with original path: {}.",
+                        arg_str,
+                        original_path
+                    );
+                })
+                .unwrap_or(arg_str) // No temp-dir prefix found, use arg as is.
+                // As the string slice we get here is a slice of memory allocated and managed by
+                // the user app, we copy the data and create new CStrings out of the copy without
+                // consuming the original data.
+                .to_owned()
+            )?
+        );
     }
+    Success(c_string_vec)
 }
 
 unsafe fn patch_sip_for_new_process(
     path: *const c_char,
     argv: *const *const c_char,
-) -> PatchedArgs {
+) -> (CString, Vec<CString>) {
     let exe_path = env::current_exe()
         .map(|path| path.to_string_lossy().to_string())
         .unwrap_or_default();
     trace!("Executable {} called execve/posix_spawn", exe_path);
 
-    // Do unsafe part of path conversion here.
-    let rawish_path = (!path.is_null()).then(|| CStr::from_ptr(path));
+    // TODO: should I change this function to return a result, and if exec/spawn are called with a
+    //       null pointer we could just pass it on instead of panicking?
+    let path_str = raw_to_str(&path).expect("Exec/Spawn hook got null pointer for path.");
 
     // Continue even if there were errors - just run without patching.
-    let path_c_string = patch_if_sip(rawish_path)
-        .success()
-        .map(CString::new)
-        .transpose()
-        .unwrap_or_default();
+    let path_c_string = patch_if_sip(path_str)
+        .and_then(|w| {
+            // TODO: is there a shorter way to convert a Result into a Detour in this situation?
+            CString::new(w).map_or_else(|err| Error(Null(err)), |c_string| Success(c_string))
```

Comment:
Is there already a good way to convert a result into a detour? If not, should I create one?

Reply:
The `?` operator should work, as long as `E: Into<HookError>`.